homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,273 @@
+
+
+DOMAIN = 'camera'
+DEPENDENCIES = ['http']
+GROUP_NAME_ALL_CAMERAS = 'all_cameras'
+SCAN_INTERVAL = 30
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+SWITCH_ACTION_RECORD = 'record'
+SWITCH_ACTION_SNAPSHOT = 'snapshot'
+
+SERVICE_CAMERA = 'camera_service'
+
+STATE_RECORDING = 'recording'
+
+DEFAULT_RECORDING_SECONDS = 30
+
+# Maps discovered services to their platforms
+DISCOVERY_PLATFORMS = {}
+
+FILE_DATETIME_FORMAT = '%Y-%m-%d_%H-%M-%S-%f'
+DIR_DATETIME_FORMAT = '%Y-%m-%d_%H-%M-%S'
+
+REC_DIR_PREFIX = 'recording-'
+REC_IMG_PREFIX = 'recording_image-'
+
+STATE_STREAMING = 'streaming'
+STATE_IDLE = 'idle'
+
+
+# pylint: disable=too-many-branches
+def setup(hass, config):
+    """ Track states and offer events for sensors. """
+
+    component = EntityComponent(
+        logging.getLogger(__name__), DOMAIN, hass, SCAN_INTERVAL,
+        DISCOVERY_PLATFORMS)
+
+    component.setup(config)
+
+    # -------------------------------------------------------------------------
+    # CAMERA COMPONENT ENDPOINTS
+    # -------------------------------------------------------------------------
+    # The following defines the endpoints for serving images from the camera
+    # via the HA http server.  This is means that you can access images from
+    # your camera outside of your LAN without the need for port forwards etc.
+
+    # Because the authentication header can't be added in image requests these
+    # endpoints are secured with session based security.
+
+    # pylint: disable=unused-argument
+    def _proxy_camera_image(handler, path_match, data):
+        """ Proxies the camera image via the HA server. """
+        entity_id = path_match.group(ATTR_ENTITY_ID)
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+            response = camera.get_camera_image()
+            handler.wfile.write(response)
+        else:
+            handler.send_response(HTTP_NOT_FOUND)
+
+    hass.http.register_path(
+        'GET',
+        re.compile(r'/api/camera_proxy/(?P<entity_id>[a-zA-Z\._0-9]+)'),
+        _proxy_camera_image,
+        require_auth=True)
+
+    # pylint: disable=unused-argument
+    def _proxy_camera_mjpeg_stream(handler, path_match, data):
+        """ Proxies the camera image as an mjpeg stream via the HA server.
+        This function takes still images from the IP camera and turns them
+        into an MJPEG stream.  This means that HA can return a live video
+        stream even with only a still image URL available.
+        """
+        entity_id = path_match.group(ATTR_ENTITY_ID)
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+
+            try:
+                camera.is_streaming = True
+                camera.update_ha_state()
+
+                handler.request.sendall(bytes('HTTP/1.1 200 OK\r\n', 'utf-8'))
```

Comment:
I did try doing it this way originally but I had some issues with mjpeg and calling end_headers.  The mjpeg format is a bit weird and requires that you send multiple headers which does not seem to play well with the end_headers method.  Doing it manually and explicitly building the headers was the only way I could get it to work.  I can tidy up the code and add in the constants you suggested though.